autoapi/toctree.py

```diff
@@ -0,0 +1,141 @@
+"""
+A small Sphinx extension that adds Domain objects (eg. Python Classes & Methods) to the TOC Tree.
+
+It dynamically adds them to the already rendered ``app.env.tocs`` dict on the Sphinx environment.
+Traditionally this only contains Section's,
+we then nest our Domain references inside the already existing Sections.
+"""
+
+from docutils import nodes
+from sphinx import addnodes
+
+
+def _build_toc_node(docname, anchor='anchor', text='test text', bullet=False):
+    """
+    Create the node structure that Sphinx expects for TOC Tree entries.
+    The ``bullet`` argument wraps it in a ``nodes.bullet_list``,
+    which is how you nest TOC Tree entries.
+    """
+
+    reference = nodes.reference('', '', internal=True, refuri=docname,
+                                anchorname='#' + anchor, *[nodes.Text(text, text)])
+    para = addnodes.compact_paragraph('', '', reference)
+    ret_list = nodes.list_item('', para)
+    if not bullet:
+        return ret_list
+    else:
+        return nodes.bullet_list('', ret_list)
+
+
+def _traverse_parent(node, objtypes):
+    """
+    Traverse up the node's parents until you hit the ``objtypes`` referenced.
+    Can either be a single type,
+    or a tuple of types.
+    """
+
+    curr_node = node.parent
+    while curr_node is not None:
+        if isinstance(curr_node, objtypes):
+            return curr_node
+        curr_node = curr_node.parent
+    return None
+
+
+def _find_toc_node(toc, ref_id, objtype):
+    """
+    Find the actual TOC node for a ref_id.
+    Depends on the object type:
+
+    * Section - First section (refuri) or 2nd+ level section (anchorname)
+    * Desc - Just use the anchor name
+    """
+    for check_node in toc.traverse(nodes.reference):
+        if objtype == nodes.section and \
+                (check_node.attributes['refuri'] == ref_id or
+                 check_node.attributes['anchorname'] == '#' + ref_id):
```

Comment:
Don't believe so, but not 100% sure. Will leave for now and see if it blows up on larger use.